tree_arena/src/tree_arena_unsafe.rs

```diff
@@ -0,0 +1,617 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(unsafe_code, reason = "Purpose is unsafe abstraction")]
+use super::NodeId;
+
+use std::cell::UnsafeCell;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+struct TreeNode<T> {
+    item: T,
+    children: Vec<NodeId>,
+}
+
+/// Mapping of data for the Tree Arena
+#[derive(Debug)]
+struct DataMap<T> {
+    /// The items in the tree
+    items: HashMap<NodeId, Box<UnsafeCell<TreeNode<T>>>>,
+    /// The parent of each node, or None if it is the root
+    parents: HashMap<NodeId, Option<NodeId>>,
+    /// The roots of the tree
+    roots: Box<UnsafeCell<Vec<NodeId>>>,
+}
+
+/// A container type for a tree of items.
+///
+/// This type is used to store zero, one or many trees of a given item type. It
+/// will keep track of parent-child relationships, lets you efficiently find
+/// an item anywhere in the tree hierarchy, and give you mutable access to this item
+/// and its children.
+#[derive(Debug)]
+pub struct TreeArena<T> {
+    /// The items in the tree
+    data_map: DataMap<T>,
+}
+
+/// A reference type giving shared access to an arena item and its children.
+///
+/// When you borrow an item from a [`TreeArena`], it returns an [`ArenaRef`].
+/// You can access it children to get access to child [`ArenaRef`] handles.
+#[derive(Debug)]
+pub struct ArenaRef<'arena, T> {
+    /// Parent of the Node
+    pub parent_id: Option<NodeId>,
+    /// Item in the node
+    pub item: &'arena T,
+    /// Children of the node
+    pub children: ArenaRefChildren<'arena, T>,
+}
+
+/// A handle giving shared access to an arena item's children.
+///
+/// See [`ArenaRef`] for more information.
+#[derive(Debug)]
+pub struct ArenaRefChildren<'arena, T> {
+    /// The associated data arena
+    parent_arena: &'arena DataMap<T>,
+    /// The parent id for these children
+    id: Option<NodeId>,
+    /// Array of children
+    #[expect(dead_code, reason = "Kept for possible future use")]
+    child_arr: &'arena [NodeId],
```

Comment:
It was kept in as I thought that it may be needed to implement tree traversal methods, but was unsure whether they would be useful - noted that https://github.com/linebender/xilem/pull/753 added an iter method over the children of a node